src/metabase/driver.clj

```diff
@@ -971,3 +974,12 @@
   {:added "0.48.0", :arglists '([driver database & args])}
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
+
+(defmulti schema+table->estimated-row-count
+  "Returns a map of [schema, table name] -> the estimated row count, for each table in the current database.
```

Comment:
I know we hold all tables in memory in a few places in sync code, but I'm wondering whether we should aim for better and instead return a reducible collection here, so we can choose not to hold everything in memory in the future.

Based on [this question on hosting insights](https://stats.metabase.com/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoyNiwidHlwZSI6InF1ZXJ5IiwicXVlcnkiOnsiYWdncmVnYXRpb24iOltbImNvdW50Il1dLCJicmVha291dCI6W1siZmllbGQiLCJpbnN0YW5jZV9pZCIseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQifV0sWyJmaWVsZCIsImRhdGFiYXNlX2lkIix7ImJhc2UtdHlwZSI6InR5cGUvSW50ZWdlciJ9XV0sIm9yZGVyLWJ5IjpbWyJkZXNjIixbImFnZ3JlZ2F0aW9uIiwwXV1dLCJzb3VyY2UtdGFibGUiOiJjYXJkX18xMzQzNCJ9fSwiZGlzcGxheSI6InRhYmxlIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9LCJvcmlnaW5hbF9jYXJkX2lkIjoxMzQzNH0=) we have some pretty large databases out there in the wild. 

I tested the memory footprint of a map similar to the one returned by `schema+table->estimated-row-count` with 1,000,000 tables, and it comes to over 200 MB. I'm pretty sure we'd prefer not to do that.

Here's my test code:
```
(require '[metabase.test :as mt])

(defn gc []
  (dotimes [_ 4] (System/gc)))

(defn used-memory []
  (let [runtime (Runtime/getRuntime)]
    (gc)
    (- (.totalMemory runtime) (.freeMemory runtime))))

(defn mbs [f]
  (let [before (used-memory)
        _ (f)
        after (used-memory)]
    (int (/ (- after before) (* 1024 1024)))))

(mbs (fn []
           (def x (into {} (take 1000000 (map vector (repeatedly (fn [] [(mt/random-name) (mt/random-name)])) (repeat 1000)))))))
=> 232 (MB)
```

Reply:
To be clear I'm not advocating against holding this data in memory now, as we're currently doing that elsewhere so this is not a memory bottleneck. I'm just suggesting to have a path to not holding the data in memory without having to change the driver method implementations.

Reply:
200 MB is really not ok, we want people (at least theoretically) to be able to run Metabase on Heroku or small EC2 instances. We need to rework this to not hold everything in memory at once. Metabase shouldn't need a gig of RAM to sync a database